File: cpp/Mod_satisfiable.cpp
#pragma once

#include "Mod_subtree.cpp"
#include "Mod_eval.cpp"

bool isVariable(String s) {
    List<Char> lst = s2lc(s);
    bool variable = false;
    char firstChar = first(lst);
    
    if (firstChar >= 'a' && firstChar <= 'z') {
        variable = true;
    }
    
    lst = removeFirst(lst);
    
    for (int i = 0; i < length(lst) ; i++) {
        Char c = get(lst, i);
        if(!((c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z') || (c >= '0' && c <= '9'))) {
            variable = false;
        }
    }
    
    return variable;
}

List<Node> getVariablesPos(Map<Node, String> map) {
    List<Node> variables = {};
    
    for (Node n: domainOf(map)) {
        if(isVariable(imageOf(map, n))) {
            variables = append(variables, n);
        }
    }
    
    return variables;
}

bool satisfiable(Tuple2<Tree, Map<Node, String>> f) {
    Map<Node, String> map = get22(f);
    List<Node> variables = getVariablesPos(map);
    
    if(empty(variables)) {
        Tuple2<Tree, Map<Node, String>> t = eval(f);
        Map<Node, String> resMap = get22(t);
        Tree resTree = get21(t);
        
        return imageOf(resMap, root(resTree)) == "TRUE" ? true : false;
    } else {
        String strue = "TRUE";
        String sflase = "FALSE";
        Map<Node, String> map1 = setImage(map, first(variables), strue);
        Map<Node, String> map2 = setImage(map, first(variables), sflase);
        
        bool b1 = satisfiable(set22(f, map1));
        bool b2 = satisfiable(set22(f, map2));

        cout << imageOf(map, first(variables)) << ";" << b1 << ";" << b2 << ";" << endl;
        
        return satisfiable(set22(f, map1)) || satisfiable(set22(f, map2));
    }
}

File: hs/Mod_satisfiable.hs
module Mod_satisfiable ( satisfiable ) where
import Prelude hiding (and, or)
import Debug.Trace
import Splib
import Types
import Mod_eval

isVariable s 
 | first s >= 'a' && first s <= 'z' = True
 | otherwise = False

getVariablesPos mapParam domain i res
 | i == length domain = res
 | isVariable (imageOf mapParam n) = getVariablesPos mapParam domain (i+1) res++[n]
 | otherwise = getVariablesPos mapParam domain (i+1) res
 where 
  n = get domain i

satisfiable ast
 | empty variables && imageOf resMap (root resTree) == "TRUE" = True
 | empty variables = False
 | otherwise = b1 || b2
 where
  evaled = eval ast
  resTree = get21 evaled
  resMap = get22 evaled
  varname = imageOf mapParam (first variables)
  map1 = setAllVariables varname mapParam (domainOf mapParam) "TRUE" 0
  map2 = setAllVariables varname mapParam (domainOf mapParam) "FALSE" 0
  b1 = satisfiable (set22 ast map1)
  b2 = satisfiable (set22 ast map2)
  variables = getVariablesPos mapParam (domainOf mapParam) 0 []
  mapParam = get22 ast

setAllVariables :: String -> (Map Node String) -> [Node] -> String -> Int -> (Map Node String)
setAllVariables name mapParam domain newV i
 | i == length domain = mapParam
 | imageOf mapParam n == name = setAllVariables name (setImage mapParam n newV) domain newV (i+1)
 | otherwise = setAllVariables name mapParam domain newV (i+1)
  where
   n = get domain i